File: 5Days-SoftwareArchitecture/typesript/Day1- Metrics/ditFragileDemo.ts
// Deep inheritance (POOR)
class Widget {
    protected enabled: boolean = true;
    render() { console.log(`Widget render (enabled=${this.enabled})`); }
}

class Button extends Widget {
    render() { console.log(`Button render (enabled=${this.enabled})`); }
}

class ActionButton extends Button {
    render() { console.log(`ActionButton render (enabled=${this.enabled})`); }
    act() { console.log("ActionButton.act()"); }
}

class SaveButton extends ActionButton {
    render() {
        if (!this.enabled) console.log("SaveButton disabled");
        else console.log("SaveButton render (saving enabled)");
    }
}

// Composition-based (GOOD)
interface IClickable { click(): void; }

class Renderer {
    render(what: string, enabled: boolean) {
        if (!enabled) console.log(`${what} render: Disabled`);
        else console.log(`${what} render: OK`);
    }
}

class PaymentHandler {
    handlePayment(amount: number) { console.log(`Handling payment: ${amount}`); }
}

class ComposeButton implements IClickable {
    private enabled = true;
    constructor(private renderer: Renderer, private paymentHandler: PaymentHandler) {}
    click(): void {
        if (!this.enabled) { console.log("ComposeButton: Disabled"); return; }
        this.renderer.render("ComposeButton", this.enabled);
        this.paymentHandler.handlePayment(99.99);
    }
    setEnabled(e: boolean) { this.enabled = e; }
}

// Demo
console.log("=== DEEP INHERITANCE (POOR) ===");
const save = new SaveButton();
console.log("Initial call:");
save.render();

// Simulate changing base-class state (fragile effect).
// In JS/TS we can mutate the inherited property directly on the instance.
(save as any).enabled = false;

console.log("After changing base 'enabled' to false (simulated base change):");
save.render();

console.log("\n=== COMPOSITION (GOOD) ===");
const renderer = new Renderer();
const payment = new PaymentHandler();
const cb = new ComposeButton(renderer, payment);
console.log("ComposeButton click (enabled true):");
cb.click();
cb.setEnabled(false);
console.log("ComposeButton click (enabled false):");
cb.click();

File: 5Days-SoftwareArchitecture/Java/Day1 CC Metrics/DitFragileDemo.java
import java.lang.reflect.Field;

// Deep inheritance (POOR)
class Widget {
    // toggling this field (default) is what can break subclasses unexpectedly
    protected boolean enabled = true;

    public void render() {
        System.out.println("Widget render (enabled=" + enabled + ")");
    }
}

class Button extends Widget {
    @Override
    public void render() {
        System.out.println("Button render (enabled=" + enabled + ")");
    }
}

class ActionButton extends Button {
    @Override
    public void render() {
        System.out.println("ActionButton render (enabled=" + enabled + ")");
    }

    public void act() {
        System.out.println("ActionButton.act()");
    }
}

class SaveButton extends ActionButton {
    @Override
    public void render() {
        // save button behavior depends on inherited 'enabled' state
        if (!enabled) System.out.println("SaveButton disabled");
        else System.out.println("SaveButton render (saving enabled)");
    }
}

// Composition-based (GOOD)
interface Clickable {
    void click();
}

class Renderer {
    void render(String what, boolean enabled) {
        if (!enabled) System.out.println(what + " render: Disabled");
        else System.out.println(what + " render: OK");
    }
}

class PaymentHandler {
    void handlePayment(double amount) {
        System.out.println("Handling payment: " + amount);
    }
}

class ComposeButton implements Clickable {
    private final Renderer renderer;
    private final PaymentHandler paymentHandler;
    private boolean enabled = true;

    ComposeButton(Renderer r, PaymentHandler p) {
        this.renderer = r;
        this.paymentHandler = p;
    }

    @Override
    public void click() {
        if (!enabled) { System.out.println("ComposeButton: Disabled"); return; }
        renderer.render("ComposeButton", enabled);
        paymentHandler.handlePayment(42.0);
    }

    public void setEnabled(boolean e) { this.enabled = e; }
}

public class DitFragileDemo {
    public static void main(String[] args) throws Exception {
        System.out.println("=== DEEP INHERITANCE (POOR) ===");
        SaveButton save = new SaveButton();
        System.out.println("Initial call:");
        save.render();

        // Simulate a change in the base class (like changing default behavior).
        // We simulate the fragile-base change by flipping the inherited field on the instance.
        // This demonstrates how a change/bug in parent-state affects descendant behavior.
        Field f = Widget.class.getDeclaredField("enabled");
        f.setAccessible(true);
        f.setBoolean(save, false); // mutate inherited field for this instance

        System.out.println("After changing base 'enabled' to false (simulated base change):");
        save.render();

        System.out.println("\n=== COMPOSITION (GOOD) ===");
        Renderer r = new Renderer();
        PaymentHandler p = new PaymentHandler();
        ComposeButton cb = new ComposeButton(r, p);
        System.out.println("ComposeButton click (enabled true):");
        cb.click();
        cb.setEnabled(false);
        System.out.println("ComposeButton click (enabled false):");
        cb.click();
    }
}
